File: python/2661.py
from typing import List

class Solution:
    def firstCompleteIndex(self, arr: List[int], mat: List[List[int]]) -> int:
        n, rows, cols = len(arr) + 1, len(mat), len(mat[0])
        n2row = [0] * n
        n2col = [0] * n
        row2F = [0] * rows
        col2F = [0] * cols
        for row in range(rows):
            for col in range(cols):
                num = mat[row][col]
                n2row[num] = row
                n2col[num] = col
        for i, num in enumerate(arr):
            row2F[n2row[num]] += 1
            col2F[n2col[num]] += 1
            if row2F[n2row[num]] == cols or col2F[n2col[num]] == rows:
                return i
        return n - 2

File: go/2661.go
package main

func firstCompleteIndex(arr []int, mat [][]int) int {
	rows := len(mat)
	cols := len(mat[0])
	n := len(arr)
	num2rows := make([]int, n+1)
	num2cols := make([]int, n+1)
	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			num2cols[mat[row][col]] = col
			num2rows[mat[row][col]] = row
		}
	}
	rowsF := make([]int, rows)
	colsF := make([]int, cols)
	for i, num := range arr {
		row := num2rows[num]
		col := num2cols[num]
		rowsF[row]++
		colsF[col]++
		if rowsF[row] == cols || colsF[col] == rows {
			return i
		}
	}
	return len(arr) - 1
}
